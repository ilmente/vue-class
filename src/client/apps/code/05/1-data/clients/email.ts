import { AxiosResponse } from 'axios';
import { netlify } from '@helpers/network-provider';
import { EmailInfo, Email, EmailStatus } from '../typings/Email';

export const getEmails = async (): Promise<EmailInfo[]> => {
    try {
        const { data }: AxiosResponse<EmailInfo[]> = await netlify({
            url: '/emails',
            method: 'GET',
        });

        return data;
    } catch (error) {
        console.error(error);
        return Promise.reject(error);
    }
}

export const getEmail = async (id: string): Promise<Email> => {
    try {
        const { data }: AxiosResponse<Email> = await netlify({
            url: '/emails',
            method: 'GET',
            params: {
                id,
            }
        });

        return data;
    } catch (error) {
        console.error(error);
        return Promise.reject(error);
    }
}

export const updateEmailStatus = async (id: string, status: EmailStatus): Promise<Email> => {
    try {
        const { data }: AxiosResponse<Email> = await netlify({
            url: '/emails',
            method: 'PATCH',
            params: {
                id,
            },
            data: {
                status
            }
        });

        return data;
    } catch (error) {
        console.error(error);
        return Promise.reject(error);
    }
}
