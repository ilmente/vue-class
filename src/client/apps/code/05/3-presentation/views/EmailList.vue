<template>
    <div class="email-list">
        <ul class="email-list__list" :class="isLoading && 'email-list__list--is-loading'">
            <li
                class="email-list__item"
                v-for="(info, index) in list" 
                :key="`email-info-${index}`">
                <router-link 
                    data-qa-email-link
                    class="email-list__link p-3"
                    active-class="has-background-white"
                    :to="{ name: RouteName.EMAIL, params: { id: info.id } }">
                    <strong class="is-size-6 has-text-grey-dark">{{ info.subject }}</strong>
                    <span class="tag is-primary is-pulled-right" :class="getStatusModifier(info.status)">
                        {{ getStatusLabel(info.status) }}
                    </span>
                    <br>
                    <a class="is-size-7">{{ info.sender }}</a>
                    <br>
                    <span class="is-size-7 has-text-grey">{{ getRelativeDate(info.timestamp) }}</span>
                </router-link>
            </li>
        </ul>
        <div v-if="isLoading" class="p-5">
            <progress 
                class="progress is-small is-primary" 
                max="100">
            </progress>
        </div>
    </div>
</template>

<style lang="scss">
    @use 'variables';

    .email-list {
        &__item {
            display: block;
            border-bottom: 1px solid variables.$grey-lightest;
        }

        &__link {
            display: block;
            line-height: 1.3;
        }

        &__list {
            will-change: opacity;
            transition: opacity 200ms linear;

            &--is-loading {
                pointer-events: none;
                opacity: 0.3;
            }
        }
    }
</style>

<script lang="ts">
    import { Component, Inject } from 'vue-property-decorator';
    import { EmailInfo } from '../../1-data/typings/Email';
    import { RouteNameType } from '../../2-connection/Router';
    import StoreComponent from '../components/StoreComponent';
    import { getRelativeDate } from '../helpers/date';
    import { getStatusLabel, getStatusModifier } from '../helpers/status';

    @Component({
        name: 'EmailList',
        methods: {
            getRelativeDate,
            getStatusLabel,
            getStatusModifier,
        }
    })
    export default class EmailList extends StoreComponent {
        @Inject() readonly RouteName!: RouteNameType;

        get list(): EmailInfo[] {
            return this.$store.getters['emailList/filteredList'];
        }

        get isLoading(): boolean {
            return this.$store.state.messaging.isLoading;
        }

        async created(): Promise<void> {
            this.$store.dispatch('emailList/loadAll');
        }
    }
</script>
