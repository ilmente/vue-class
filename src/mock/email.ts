import { generateEmails } from '@lambda/emails';
import omit from 'lodash/omit';
import { Email, EmailInfo, EmailStatus } from '../client/apps/code/05/1-data/typings/Email';

export const infoArray: EmailInfo[] = generateEmails(3).map((email: Email) => omit(email, 'content'));
export const singleEmail: Email = generateEmails(1)[0];

export const staticInfoArray: EmailInfo[] = [
    {
        id: 'f0d2b891-7f07-485a-abc0-08280fc24bac',
        sender: 'Nikki_Brekke@example.org',
        subject: 'Gorgeous Rubber Chicken',
        timestamp: 'Wed Mar 03 2021 20:30:50 GMT+0100 (Central European Standard Time)',
        status: EmailStatus.READ,
    },
    {
        id: 'e42be0fb-e63e-450c-9de5-f0f1c2367a85',
        sender: 'Emmanuel_Lemke65@example.org',
        subject: 'Handcrafted Rubber Salad',
        timestamp: 'Wed Mar 03 2021 12:38:01 GMT+0100 (Central European Standard Time)',
        status: EmailStatus.UNREAD,
    },
    {
        id: 'a7aca6ec-9b50-4e5a-93cf-7019b0b53859',
        sender: 'Daphney_Ledner13@example.com',
        subject: 'Small Soft Bike',
        timestamp: 'Wed Mar 03 2021 20:14:57 GMT+0100 (Central European Standard Time)',
        status: EmailStatus.SPAM,
    }
];

export const staticSingleEmail: Email = {
    id: 'f0d2b891-7f07-485a-abc0-08280fc24bac',
    sender: 'Nikki_Brekke@example.org',
    subject: 'Gorgeous Rubber Chicken',
    timestamp: 'Wed Mar 03 2021 20:30:50 GMT+0100 (Central European Standard Time)',
    content: 'Labore perspiciatis adipisci accusantium maiores quia sed. Et illum dolor. Possimus itaque ex reiciendis occaecati veniam. Odio dolorum veniam quidem. Tenetur quibusdam est dolor perspiciatis. Rerum neque ut non.\n' +
        ' \rConsequatur quas est ipsam mollitia harum doloremque fugiat expedita. Impedit culpa dolorum occaecati. Culpa quaerat qui cupiditate et voluptatum esse delectus.\n' +
        ' \rNisi alias itaque deleniti possimus ut necessitatibus ratione molestiae vero. Quaerat sapiente voluptas officiis. Nulla eos dolorem voluptatem beatae est. Expedita illum ea impedit corporis totam reprehenderit et cum officiis.',
    status: EmailStatus.READ,
};
