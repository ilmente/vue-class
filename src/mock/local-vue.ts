import { jest } from '@jest/globals';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import Vuex, { StoreOptions } from 'vuex';
import { EmailListState } from '@05/2-connection/modules/EmailListState';
import { EmailState } from '@05/2-connection/modules/EmailState';
import { MessagingState } from '@05/2-connection/modules/MessagingState';
import { staticInfoArray, staticSingleEmail } from './email';
import merge from 'lodash/merge';

export const localVueWithRouter = createLocalVue();
localVueWithRouter.use(VueRouter);
localVueWithRouter.use(Vuex);

export const localVue = createLocalVue();
localVue.use(Vuex);

export const router = new VueRouter();

export const generateStoreContext = (customOptions: StoreOptions<unknown> = {}) => {
    const defaultOptions = {
        modules: {
            emailList: {
                namespaced: true,
                state: new EmailListState,
                getters: {
                    filteredList: () => staticInfoArray,
                },
                actions: {
                    loadAll: jest.fn(),
                    setSpamOnly: jest.fn(),
                },
            },
            email: {
                namespaced: true,
                state: {
                    ...new EmailState,
                    current: staticSingleEmail,
                },
                getters: {
                    exists: () => true,
                    content: () => 'Content',
                },
                actions: {
                    load: jest.fn(),
                    updateEmailStatus: jest.fn(),
                },
            },
            messaging: {
                state: new MessagingState,
                actions: {
                    setLoading: jest.fn(),
                    push: jest.fn(),
                },
            }
        }
    };

    const options = merge(defaultOptions, customOptions);

    return {
        store: new Vuex.Store(options),
        options,
    }
};
