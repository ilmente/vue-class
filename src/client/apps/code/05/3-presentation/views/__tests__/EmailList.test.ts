import { mount } from '@vue/test-utils'
import EmailList from '../EmailList.vue';
import { localVueWithRouter, router } from '@test/local-vue';
import { staticInfoArray } from '@mock/email';
import Vuex from 'vuex';

const provide = {
    RouteName: { EMAIL: '#' }
}

describe('EmailList.vue', () => {
    let store: any;
    let state: any;
    let actions: any;
    let getters: any;

    const getWrapper = () => mount(EmailList, {
        localVue: localVueWithRouter,
        router,
        store,
        provide,
        stubs: ['router-link'],
    });

    beforeEach(() => {
        state = {
            isLoading: false,
        }

        getters = {
            filteredList: () => staticInfoArray,
        }

        actions = {
            loadAll: jest.fn()
        }

        store = new Vuex.Store({
            modules: {
                emailList: {
                    actions,
                    getters,
                    namespaced: true
                },
                messaging: {
                    state,
                }
            }
        })
    });

    test('snapshot', () => {
        const wrapper = getWrapper();
        expect(wrapper.element).toMatchSnapshot();
    });

    test('call loadAll', () => {
        getWrapper();
        expect(actions.loadAll).toHaveBeenCalledTimes(1);
    });
});
