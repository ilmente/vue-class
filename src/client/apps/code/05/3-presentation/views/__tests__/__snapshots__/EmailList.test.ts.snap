// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailList.vue snapshot 1`] = `
<div
  class="email-list"
>
  <ul
    class="email-list__list"
  >
    <li
      class="email-list__item"
    >
      <router-link-stub
        activeclass="has-background-white"
        ariacurrentvalue="page"
        class="email-list__link p-3"
        data-qa-email-link=""
        event="click"
        tag="a"
        to="[object Object]"
      >
        <strong
          class="is-size-6 has-text-grey-dark"
        >
          Gorgeous Rubber Chicken
        </strong>
         
        <span
          class="tag is-primary is-pulled-right is-info is-light"
        >
          
                    Read
                
        </span>
         
        <br />
         
        <a
          class="is-size-7"
        >
          Nikki_Brekke@example.org
        </a>
         
        <br />
         
        <span
          class="is-size-7 has-text-grey"
        >
          yesterday at 8:30 PM
        </span>
      </router-link-stub>
    </li>
    <li
      class="email-list__item"
    >
      <router-link-stub
        activeclass="has-background-white"
        ariacurrentvalue="page"
        class="email-list__link p-3"
        data-qa-email-link=""
        event="click"
        tag="a"
        to="[object Object]"
      >
        <strong
          class="is-size-6 has-text-grey-dark"
        >
          Handcrafted Rubber Salad
        </strong>
         
        <span
          class="tag is-primary is-pulled-right is-success is-light"
        >
          
                    Unread
                
        </span>
         
        <br />
         
        <a
          class="is-size-7"
        >
          Emmanuel_Lemke65@example.org
        </a>
         
        <br />
         
        <span
          class="is-size-7 has-text-grey"
        >
          yesterday at 12:38 PM
        </span>
      </router-link-stub>
    </li>
    <li
      class="email-list__item"
    >
      <router-link-stub
        activeclass="has-background-white"
        ariacurrentvalue="page"
        class="email-list__link p-3"
        data-qa-email-link=""
        event="click"
        tag="a"
        to="[object Object]"
      >
        <strong
          class="is-size-6 has-text-grey-dark"
        >
          Small Soft Bike
        </strong>
         
        <span
          class="tag is-primary is-pulled-right is-danger is-light"
        >
          
                    Spam
                
        </span>
         
        <br />
         
        <a
          class="is-size-7"
        >
          Daphney_Ledner13@example.com
        </a>
         
        <br />
         
        <span
          class="is-size-7 has-text-grey"
        >
          yesterday at 8:14 PM
        </span>
      </router-link-stub>
    </li>
  </ul>
   
  <!---->
</div>
`;
